File: src/inter/constant.cpp
#include <inter/constant.hpp>

#include <lexer/word.hpp>

namespace inter {
std::shared_ptr<Constant> Constant::kTrue = Constant::create(lexer::Token::create(lexer::Word::kTrue),
                                                             symbols::Type::boolean);
std::shared_ptr<Constant> Constant::kFalse = Constant::create(lexer::Token::create(lexer::Word::kFalse),
                                                              symbols::Type::boolean);
} // namespace inter

File: java/com/dragon/inter/Constant.java
package com.dragon.inter;

import com.dragon.lexer.Num;
import com.dragon.lexer.Token;
import com.dragon.lexer.Word;
import com.dragon.symbols.Type;

public class Constant extends Expr {
  public static final Constant True = new Constant(Word.True, Type.Bool);
  public static final Constant False = new Constant(Word.False, Type.Bool);

  public Constant(Token tok, Type p) {
    super(tok, p);
  }

  public Constant(int i) {
    super(new Num(i), Type.Int);
  }

  @Override
  public void jumping(int t, int f) {
    if (this == True && t != 0) {
      emit("goto L" + t);
    } else if (this == False && f != 0) {
      emit("goto L" + f);
    }
  }
}
